### ISSUE TITLE
`once` places the return value into unwanted Scalar container

### ISSUE THREAD

**Issue by zoffixznet** [rakudo/rakudo#1610]
```
m: (once 42).VAR.^name.say
rakudo-moar 52e66ad30: OUTPUT: «Scalar␤»
```

Which produces unexpected behaviour like this:
```
m: my \z := once 42; z = 100; say z
rakudo-moar 52e66ad30: OUTPUT: «100␤»
```

**Comment by lizmat** [comment]
Guess we need some change near line 2359 in src/Perl6/Actions.nqp

**Comment by zoffixznet** [comment]
Looks like it's handled with a state var. Not sure if it's by-design, but looks like binding to statevars doesn't work:

```
    $ perl6 -e 'for ^1 { state $x := 42; say $x }'
    (Mu)
```

Otherwise, I imagine changing [this `p6store`](https://github.com/rakudo/rakudo/blob/52e66ad30c05d291cd66dd84b5b8bded04c3d11d/src/Perl6/Actions.nqp#L2363) to `bind` would do the trick.

**Comment by zoffixznet** [comment]
>  it's handled with a state var

Or maybe not... I see `p6stateinit` that isn't documented, but the QAST::Var itself is declared with `scope<lexical>`

**Comment by lizmat** [comment]
The statevar is invisible: it’s there to ensure the “once” only happens once.

This was one of the first things I did in the core.  I wonder if this could not be done with a cas() nowadays.

**Comment by lizmat** [comment]
I guess not since: "A block prefix with once will be executed exactly once, even if placed inside a loop or a recursive routine."

**Comment by lizmat** [comment]
`p6stateinit` isn't document because it is a Rakudo specific nqp op that lives in src/vm/moar/ops/perl6_ops.c

**Comment by lizmat** [comment]
Fixed with f97398cfb8

**Comment by zoffixznet** [comment]
> p6stateinit isn't document because it is a Rakudo specific nqp op

We have those documented in https://github.com/rakudo/rakudo/blob/master/docs/ops.markdown but `p6stateinit` is merely listed, not described.
